### ISSUE TITLE
Return type from a route is not being checked against the schema, 

### ISSUE THREAD

**Issue by respondsid** [lukeautry/tsoa#808]
When a Promise is retuened from a @service (in my case its a js service being called from ts controller, js service returns an object which adheres to the return type of the ts controller). When I look at the generated route.ts file aparently no ref is generated for return type and a global handler is generated which simply flushes out the output res.json(data)

## Sorting

- **I'm submitting a ...**

  - [ *] bug report
  - [ ] feature request
  - [ ] support request

- I confirm that I
  - [ ] used the [search](https://github.com/lukeautry/tsoa/search?type=Issues) to make sure that a similar issue hasn't already been submit

## Expected Behavior
prior to 
 res.json(data)
the generated TS code should enforce the return type contract and reduce the returned object from service (data) to the  ts contract 

```yaml

```

```ts
```

## Current Behavior
Currently if a controller returns a Promise of any specific type, and from service if we return a different type (Considering service is a js file), the tsoa module doesnt enforce the outgoing contract on the json object return as part of the route invocation, note that it does perform type based validations in request body.

```yaml

```

```ts
```

## Possible Solution

In route generator template the promiseHandler should also consider ref: of response type and should try and reduce the service retuened object (data) to the returned type interface.
## Steps to Reproduce

1. 2. 3. 4.

## Context (Environment)

Version of the library: 
Version of NodeJS: 

- Confirm you were using yarn not npm: [ ]

**Comment by lukeautry** [comment]
This is a good call out and I can see how this would be expected behavior.

The route file already has the information available to validate an object, so I think it's not a huge lift to do validation on the response.

However, validating responses can lead to problems and poor performance. I've worked on APIs where the response contained an array with over 100,000 elements, so running validation on this entire result set will have a non-negligible impact on response time.

I can think of a few paths forward:
- Status quo: don't do response validation
- Progressive enhancement: give tsoa users an option of doing response validation, if it's important to them and if they're willing to deal with the performance impact, whatever the impact may be
- Do response validation for all tsoa users
  - I don't think this is a great option; this could easily result in regressions and breakage in production apps. I'd prefer to not do this.

thoughts @WoH?

**Comment by WoH** [comment]
What would you do you do if the response does not validate?
I am opposed to enabling this behaviour by default, but I could see someone turning this on in development. I'm pretty convinced TS is a better tool to enforce types, but the info is there, may aswell use it.

**Comment by respondsid** [comment]
So my understanding is that it should throw a type casting error. With the current implementation the type is resolved by grnerated route js. So they type enforcement aldo needs to be therr . I agrre that it csn be configurable however i believe a type based framework should have sn option to enforce typed in/out

**Comment by lukeautry** [comment]
> What would you do you do if the response does not validate?

Yeah, this is a big issue IMO. As an API client, you care that your request is invalid. You don't care that the developer-defined response is invalid; you just want some response to work with.

In cases where you control the server and the client, sure, you may want the request to blow up to preserve the contract. I really think this is something you'd only want in dev, not prod.

**Comment by respondsid** [comment]
If TSOA has an option to define a response as  Promise<T>, I think tsoa should ate least look at T and do an instance of check ? before returning and should blow up with a standard error so that it gets logged and developer gets a notification. If we dont do that and send a completely difference response R our consumer will anyway  fail even worst we will only get to know about it when a consumer contacts and tells us (Given that we don't have an automated test covering this).

**Comment by WoH** [comment]
I really think you should take a look at using (strict) TS to check that before runtime.
A response r where r is not of shape T >>>>>>>> no response.
The instance check will only work on Classes, which would only work for a very small subset of possible responses.

**Comment by vpillinger** [comment]
Response validation could be very worthwhile for development purposes. For a trivial example one would want to make sure to not sending "password" field back in a response which it shouldn't be returned in.

Although, it may also make more sense to use a tool like [Dredd](https://dredd.org/en/latest/) to run integration tests rather than being done at runtime.

**Comment by WoH** [comment]
How would that not be covered by the return type of the controller method?

**Comment by vpillinger** [comment]
In my case I want to define public interfaces for my classes in regards to API, but I want to have additional properties in my internal model. For example, I want to ensure that I always drop "password" when I translate my output to the API, but I could foresee-ably do the following:

```
// interface definition
export interface UserAPI {
	readonly id: number;
	email: string;
	name: string;
	status?: "Happy" | "Sad";
	phoneNumbers: string[];
}

export interface User {
	readonly id: number;
	email: string;
	name: string;
	status?: "Happy" | "Sad";
	phoneNumbers: string[];
}

export class AdminUserReal implements User {
	id: number;
	password: string;
	email: string;
	name: string;
	status?: "Happy" | "Sad";
	phoneNumbers: string[];
	adminAccessId: number;

	constructor() {
		this.id = 1;
		this.password = "hashedPassword";
		this.email = "foo@bar.com";
		this.name = "foo barino";
		this.status = "Happy";
		this.phoneNumbers = []
		this.adminAccessId = 1234;
	}

	toAPI(): User {
		return this;
	}
}
```

**Comment by WoH** [comment]
So the return type is the interface which will lead to an error if you return the class or additional properties which are not permitted by the interface.
It's a generally advisable concept to split domain objects and classes from API interfaces (and start with the API interfaces first).
While we support classes it's always advisable to think about the API contract via interfaces first.

I don't really understand the argument for runtime checks given the examples you gave after all.

**Comment by WoH** [comment]
Thanks for the use case which helps me a lot in understanding the narrowed down issue.
Correct me if I'm wrong, but the narrowed down feature request boils down to:
The structural typing TS provides around `this` by default has undesirable properties here because it does not check for excess properties (as strict as we'd like here), a check you would like to have in order to ensure that no excess properties are returned with regards to the return type.
I would certainly agree that the default behavior is not the ideal tool here, but, for the reasons mentioned previously, I tend to think runtime checking may not be the best approach either.

Instead, I think a helper may be the best middle ground.

```ts
type Exact<T, S> =
  T extends S ? 
  Exclude<keyof T, keyof S> extends never ? 
 T : never : never;


function safe<S, T>(check: Exact<T, S>): S {
	return check;
}
```

[Playground](https://www.typescriptlang.org/play?noUnusedLocals=true&noUnusedParameters=true&target=7&jsx=0&ssl=39&ssc=38&pln=39&pc=51#code/PTAEEEAUElQSwHYBcCmAnAZgQwMYtEgBZZKiwDuWyoGA9mqALYCeEMoAJrTgM4BQiVJlz4Aqj3RRYAbz4BINCixcEAG1ZwOALlAIArowBG6ANzyUjLHFU6eSNIgDmZuQiyMUt+05d2SengB+HQAiAAksAAdI5hDQAB9QEIBlZRCXSMJaBBQAOQNjNB4vBwRHAG0AXTMAXz4+HFUsHh5QcXR4RkjVCxRkVva0KVBZOU0dfSNTeUjmnnJ6bVA7UudzS2sSn3k3Dy2y3yR-INCI6NiEpNSOdJmsnPypov2K6vllRkRwHDwW6CXJoUzPIcNkVnocEh6AAKACUI3kciIcB4ADpNKAALygACMLiRhBRqNmLQWaA4WKSxB4hBQHEgczJN3xyLRFisqkpITotAAAoYsGhUaDGLc5ASibt8NjubRaKABaVaGKJWi-EgAlyzjEVazifc8gV0K1sVVEXqPl8fig-hTsTiAEwAZgALC46vUkbQpAAVQk8OE6QbDUYKFAatAIZZYDAoaGs2Huz1Q33+gAiNqQgbaEiG7FDigjUZ4Mbj0lRFdZNUT8g9XqkAHU4EQAKIAD1+AdhQdzIcRoIQdhGmgANOzrCOpSPMtlDU8eCPQJaEN9O-8apTWfjEYW9JHo7HodJR6Bx6pF1PQDOHkaiovl6ubTx1zW5HWUzBs8H8-2waRjxwY4bOeU7XnOhQ8Bu2JboiO7hnuxalkeJ5nhe7goIuYGPBB1ZJh6SDMJE+DtrgSAADw+ouyQAHxYnwoCgD6p5tqgCAcK0ySgIEoD0aA7aNHoHAoGRADWKDMLQGCMYuYkSVJNHMax7G6CgABuHTcfRTETGpHQ6epaDAnwGB6AgkJwNkB7Ccki4+tR0I4LSOAiToJGQhRVHUd2oCcaMu77o5KDObU9QNGCtA9Kiqi0I40I5OQOboHCqIftAcKwkAA)

Wdyt?

**Comment by MatthiasKunnen** [comment]
@WoH would you be open to adding an option akin to `noImplicitAdditionalProperties `?

**Comment by WoH** [comment]
My views on this have (somewhat) shifted. While my concern initially was the performance, I have found approaches that use JSON Schemas to, in fact, improve serialization. I've wanted to reconsider this issue, but I don't have a lot of time to spare for experiments here, but if someone wants to look at fastify for inspiration and propose a similar solution, I'm all for that.

**Comment by blipk** [comment]
I think this is better left up to other libraries such as `class-validator` or `zod` or whatever for validating responses - or perhaps tighter integration with `tsoa` and `class-validator` as they both have a nice decorator approach.

If your code is properly typed it should get you most of the way there for validating responses, and then the libraries I mentioned can be used.

I also suggest looking at FastAPI + Pydantic (+ SQLAlchemy) implementations for how a good OpenAPI-as-code should work.
